// hyperformula/commonjs/Evaluator.js
"use strict";

require("core-js/modules/es.array.slice.js");

require("core-js/modules/es.function.name.js");

require("core-js/modules/es.array.from.js");

require("core-js/modules/es.string.iterator.js");

require("core-js/modules/es.regexp.exec.js");

require("core-js/modules/es.symbol.js");

require("core-js/modules/es.symbol.description.js");

require("core-js/modules/es.symbol.iterator.js");

require("core-js/modules/es.array.iterator.js");

require("core-js/modules/web.dom-collections.iterator.js");

exports.__esModule = true;
exports.Evaluator = void 0;

require("core-js/modules/es.object.to-string.js");

require("core-js/modules/web.dom-collections.for-each.js");

var _AbsoluteCellRange = require("./AbsoluteCellRange");

var _absolutizeDependencies = require("./absolutizeDependencies");

var _Cell = require("./Cell");

var _ContentChanges = require("./ContentChanges");

var _DependencyGraph = require("./DependencyGraph");

var _FormulaCellVertex = require("./DependencyGraph/FormulaCellVertex");

var _InterpreterState = require("./interpreter/InterpreterState");

var _InterpreterValue = require("./interpreter/InterpreterValue");

var _SimpleRangeValue = require("./interpreter/SimpleRangeValue");

var _statistics = require("./statistics");

function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it.return != null) it.return(); } finally { if (didErr) throw err; } } }; }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }

var Evaluator = /*#__PURE__*/function () {
  function Evaluator(config, stats, interpreter, lazilyTransformingAstService, dependencyGraph, columnSearch) {
    _classCallCheck(this, Evaluator);

    this.config = config;
    this.stats = stats;
    this.interpreter = interpreter;
    this.lazilyTransformingAstService = lazilyTransformingAstService;
    this.dependencyGraph = dependencyGraph;
    this.columnSearch = columnSearch;
  }

  _createClass(Evaluator, [{
    key: "run",
    value: function run() {
      var _this = this;

      this.stats.start(_statistics.StatType.TOP_SORT);

      var _this$dependencyGraph = this.dependencyGraph.topSortWithScc(),
          sorted = _this$dependencyGraph.sorted,
          cycled = _this$dependencyGraph.cycled;

      this.stats.end(_statistics.StatType.TOP_SORT);
      this.stats.measure(_statistics.StatType.EVALUATION, function () {
        _this.recomputeFormulas(cycled, sorted);
      });
    }
  }, {
    key: "partialRun",
    value: function partialRun(vertices) {
      var _this2 = this;

      var changes = _ContentChanges.ContentChanges.empty();

      this.stats.measure(_statistics.StatType.EVALUATION, function () {
        _this2.dependencyGraph.graph.getTopSortedWithSccSubgraphFrom(vertices, function (vertex) {
          if (vertex instanceof _FormulaCellVertex.FormulaVertex) {
            var currentValue = vertex.isComputed() ? vertex.getCellValue() : undefined;

            var newCellValue = _this2.recomputeFormulaVertexValue(vertex);

            if (newCellValue !== currentValue) {
              var address = vertex.getAddress(_this2.lazilyTransformingAstService);
              changes.addChange(newCellValue, address);

              _this2.columnSearch.change((0, _InterpreterValue.getRawValue)(currentValue), (0, _InterpreterValue.getRawValue)(newCellValue), address);

              return true;
            }

            return false;
          } else if (vertex instanceof _DependencyGraph.RangeVertex) {
            vertex.clearCache();
            return true;
          } else {
            return true;
          }
        }, function (vertex) {
          if (vertex instanceof _DependencyGraph.RangeVertex) {
            vertex.clearCache();
          } else if (vertex instanceof _FormulaCellVertex.FormulaVertex) {
            var address = vertex.getAddress(_this2.lazilyTransformingAstService);

            _this2.columnSearch.remove((0, _InterpreterValue.getRawValue)(vertex.valueOrUndef()), address);

            var error = new _Cell.CellError(_Cell.ErrorType.CYCLE, undefined, vertex);
            vertex.setCellValue(error);
            changes.addChange(error, address);
          }
        });
      });
      return changes;
    }
  }, {
    key: "runAndForget",
    value: function runAndForget(ast, address, dependencies) {
      var _this3 = this;

      var tmpRanges = [];

      var _iterator = _createForOfIteratorHelper((0, _absolutizeDependencies.absolutizeDependencies)(dependencies, address)),
          _step;

      try {
        for (_iterator.s(); !(_step = _iterator.n()).done;) {
          var dep = _step.value;

          if (dep instanceof _AbsoluteCellRange.AbsoluteCellRange) {
            var range = dep;

            if (this.dependencyGraph.getRange(range.start, range.end) === undefined) {
              var rangeVertex = new _DependencyGraph.RangeVertex(range);
              this.dependencyGraph.rangeMapping.setRange(rangeVertex);
              tmpRanges.push(rangeVertex);
            }
          }
        }
      } catch (err) {
        _iterator.e(err);
      } finally {
        _iterator.f();
      }

      var ret = this.evaluateAstToCellValue(ast, new _InterpreterState.InterpreterState(address, this.config.useArrayArithmetic));
      tmpRanges.forEach(function (rangeVertex) {
        _this3.dependencyGraph.rangeMapping.removeRange(rangeVertex);
      });
      return ret;
    }
    /**
     * Recalculates formulas in the topological sort order
     */

  }, {
    key: "recomputeFormulas",
    value: function recomputeFormulas(cycled, sorted) {
      var _this4 = this;

      cycled.forEach(function (vertex) {
        if (vertex instanceof _FormulaCellVertex.FormulaVertex) {
          vertex.setCellValue(new _Cell.CellError(_Cell.ErrorType.CYCLE, undefined, vertex));
        }
      });
      sorted.forEach(function (vertex) {
        if (vertex instanceof _FormulaCellVertex.FormulaVertex) {
          var newCellValue = _this4.recomputeFormulaVertexValue(vertex);

          var address = vertex.getAddress(_this4.lazilyTransformingAstService);

          _this4.columnSearch.add((0, _InterpreterValue.getRawValue)(newCellValue), address);
        } else if (vertex instanceof _DependencyGraph.RangeVertex) {
          vertex.clearCache();
        }
      });
    }
  }, {
    key: "recomputeFormulaVertexValue",
    value: function recomputeFormulaVertexValue(vertex) {
      var address = vertex.getAddress(this.lazilyTransformingAstService);

      if (vertex instanceof _DependencyGraph.ArrayVertex && (vertex.array.size.isRef || !this.dependencyGraph.isThereSpaceForArray(vertex))) {
        return vertex.setNoSpace();
      } else {
        var formula = vertex.getFormula(this.lazilyTransformingAstService);
        var newCellValue = this.evaluateAstToCellValue(formula, new _InterpreterState.InterpreterState(address, this.config.useArrayArithmetic, vertex));
        return vertex.setCellValue(newCellValue);
      }
    }
  }, {
    key: "evaluateAstToCellValue",
    value: function evaluateAstToCellValue(ast, state) {
      var interpreterValue = this.interpreter.evaluateAst(ast, state);

      if (interpreterValue instanceof _SimpleRangeValue.SimpleRangeValue) {
        return interpreterValue;
      } else if (interpreterValue === _InterpreterValue.EmptyValue && this.config.evaluateNullToZero) {
        return 0;
      } else {
        return interpreterValue;
      }
    }
  }]);

  return Evaluator;
}();

exports.Evaluator = Evaluator;
